policy-engine/src/graph.rs

```diff
@@ -18,20 +18,49 @@ use cincinnati::{Graph, CONTENT_TYPE};
 use failure::Error;
 use futures::{future, Future, Stream};
 use hyper::{Body, Client, Request, Uri};
+use prometheus::{Counter, Histogram};
 use serde_json;
 
-pub fn index(req: HttpRequest<State>) -> Box<Future<Item = HttpResponse, Error = Error>> {
+lazy_static! {
+    static ref HTTP_GRAPH_REQS: Counter = register_counter!(
+        "http_graph_requests_total",
+        "Total number of HTTP /v1/graph requests."
+    ).unwrap();
+    static ref HTTP_GRAPH_WRONG_REQS: Counter = register_counter!(
+        "http_graph_wrong_requests_total",
+        "Total number of wrong HTTP /v1/graph requests."
+    ).unwrap();
+    static ref HTTP_UPSTREAM_REQS: Counter = register_counter!(
+        "http_upstream_requests_total",
+        "Total number of HTTP upstream requests."
+    ).unwrap();
+    static ref HTTP_UPSTREAM_UNREACHABLE: Counter = register_counter!(
+        "http_upstream_errors_total",
+        "Total number of HTTP upstream unreachable errors."
+    ).unwrap();
+    static ref HTTP_SERVE_HIST: Histogram = register_histogram!(
+        "http_graph_serve_duration_seconds",
+        "HTTP graph serving latency in seconds."
+    ).unwrap();
+}
+
+pub(crate) fn index(req: HttpRequest<State>) -> Box<Future<Item = HttpResponse, Error = Error>> {
+    HTTP_GRAPH_REQS.inc();
     match req.headers().get(header::ACCEPT) {
-        Some(entry) if entry == HeaderValue::from_static(CONTENT_TYPE) => Box::new(
-            Client::new()
-                .request(
-                    Request::get(&req.state().upstream)
-                        .header(header::ACCEPT, HeaderValue::from_static(CONTENT_TYPE))
-                        .body(Body::empty())
-                        .expect("unable to form request"),
-                )
+        Some(entry) if entry == HeaderValue::from_static(CONTENT_TYPE) => {
+            let ups_req = Request::get(&req.state().upstream)
+                .header(header::ACCEPT, HeaderValue::from_static(CONTENT_TYPE))
+                .body(Body::empty())
+                .expect("unable to form request");
+            HTTP_UPSTREAM_REQS.inc();
+            let timer = HTTP_SERVE_HIST.start_timer();
+            let serve = Client::new()
+                .request(ups_req)
                 .from_err::<Error>()
-                .and_then(|res| {
+                .map_err(|e| {
+                    HTTP_UPSTREAM_UNREACHABLE.inc();
+                    e
+                }).and_then(|res| {
```

Comment:
Can you bump this down to the line below? This formatting is a bug in rustfmt that should have been fixed recently.

Reply:
Did a pass with nightly rustfmt, moved a few other things as well, it looks better now.